crates/win-api-wrappers/src/wts.rs

```diff
@@ -87,24 +98,232 @@ impl Drop for WTSVirtualChannel {
+}
+
+impl TryFrom<WTS_CONNECTSTATE_CLASS> for WTSConnectState {
+    type Error = anyhow::Error;
+
+    fn try_from(v: WTS_CONNECTSTATE_CLASS) -> Result<Self, Self::Error> {
+        match v {
+            windows::Win32::System::RemoteDesktop::WTSActive => Ok(WTSConnectState::Active),
+            windows::Win32::System::RemoteDesktop::WTSConnected => Ok(WTSConnectState::Connected),
+            windows::Win32::System::RemoteDesktop::WTSConnectQuery => Ok(WTSConnectState::ConnectQuery),
+            windows::Win32::System::RemoteDesktop::WTSShadow => Ok(WTSConnectState::Shadow),
+            windows::Win32::System::RemoteDesktop::WTSDisconnected => Ok(WTSConnectState::Disconnected),
+            windows::Win32::System::RemoteDesktop::WTSIdle => Ok(WTSConnectState::Idle),
+            windows::Win32::System::RemoteDesktop::WTSListen => Ok(WTSConnectState::Listen),
+            windows::Win32::System::RemoteDesktop::WTSReset => Ok(WTSConnectState::Reset),
+            windows::Win32::System::RemoteDesktop::WTSDown => Ok(WTSConnectState::Down),
+            windows::Win32::System::RemoteDesktop::WTSInit => Ok(WTSConnectState::Init),
+            _ => Err(anyhow::anyhow!("Invalid WTS_CONNECTSTATE_CLASS: {}", v.0)),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct WTSSessionInfo {
+    pub session_id: u32,
+    pub win_station_name: String,
+    pub state: WTSConnectState,
+}
+
+impl WTSSessionInfo {
+    /// # Safety
+    ///
+    /// - The `pWinStationName` pointer needs to be valid for reads up until and including the next `\0`.
+    unsafe fn new(wts_session: &WTS_SESSION_INFOW) -> anyhow::Result<Self> {
+        Ok(Self {
+            session_id: wts_session.SessionId,
+            win_station_name: unsafe { wts_session.pWinStationName.to_string()? },
+            state: wts_session.State.try_into()?,
+        })
+    }
+}
+
+pub fn get_sessions() -> anyhow::Result<Vec<WTSSessionInfo>> {
+    let mut sessions_ptr: *mut WTS_SESSION_INFOW = std::ptr::null_mut();
+    let mut len = 0u32;
+
+    // SAFETY: All buffers are valid.
+    unsafe {
+        WTSEnumerateSessionsW(WTS_CURRENT_SERVER_HANDLE, 0, 1, &mut sessions_ptr, &mut len)?;
+    };
+
+    // SAFETY: `sessions_ptr` is always a valid pointer on success.
+    // `sessions_ptr` must be freed by `WTSFreeMemory`.
+    let sessions_ptr = unsafe { WTSMemory::from_raw(sessions_ptr.cast()) };
+
+    // SAFETY: Verify that all the safety preconditions of from_raw_parts are uphold: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety
+    let sessions_slice: &[WTS_SESSION_INFOW] =
+        unsafe { std::slice::from_raw_parts(sessions_ptr.0 as *mut u8 as *mut WTS_SESSION_INFOW, len as usize) };
+
+    let mut sessions = Vec::<WTSSessionInfo>::with_capacity(sessions_slice.len());
+    for session in sessions_slice {
+        // SAFETY: `session` is valid
+        unsafe {
+            sessions.push(WTSSessionInfo::new(session)?);
+        }
+    }
```

Comment:
suggestion: It’s considered good style to shadow variables that should not be used anymore, especially when you can use this to prove correctness of the code:

```rust
// SAFETY:
// - `WTSEnumerateSessionsW` returns a pointer valid for `len` reads of `WTS_SESSION_INFOW`.
// - `WTSEnumerateSessionsW` is also returning a pointer properly aligned.
// - We ensure the memory referenced by the slice is not mutated by shadowing the variable.
// - There are never so many sessions that `len * mem::size_of::<WTS_SESSION_INFOW>()` overflows `isize`.
let sessions: &[WTS_SESSION_INFOW] =
    unsafe { std::slice::from_raw_parts(sessions.0 as *mut u8 as *mut WTS_SESSION_INFOW, len as usize) };
```

Here is an improved version of this code:

```suggestion
    let mut sessions: *mut WTS_SESSION_INFOW = std::ptr::null_mut();
    let mut len = 0u32;

    // SAFETY: All buffers are valid.
    unsafe {
        WTSEnumerateSessionsW(WTS_CURRENT_SERVER_HANDLE, 0, 1, &mut sessions, &mut len)?;
    };

    // SAFETY: `sessions` is always a valid pointer on success.
    // `sessions` must be freed by `WTSFreeMemory`.
    let sessions = unsafe { WTSMemory::from_raw(sessions.cast()) };

    // SAFETY:
    // - `WTSEnumerateSessionsW` returns a pointer valid for `len` reads of `WTS_SESSION_INFOW`.
    // - `WTSEnumerateSessionsW` is also returning a pointer properly aligned.
    // - We ensure the memory referenced by the slice is not mutated by shadowing the variable.
    // - There are never so many sessions that `len * mem::size_of::<WTS_SESSION_INFOW>()` overflows `isize`.
    let sessions: &[WTS_SESSION_INFOW] =
        unsafe { std::slice::from_raw_parts(sessions.0 as *mut u8 as *mut WTS_SESSION_INFOW, len as usize) };

    let sessions = sessions.iter()
        .map(|session| {
            // SAFETY: `session` is a WTS_SESSION_INFOW initialized by Remote Desktop Services that we trust to be valid.
            unsafe { WTSSessionInfo::new(session) }
        })
        .collect::<Result<Vec<WTSSessionInfo>, _>>()?;
```